### PR TITLE
Fix parsing of error in fetch tool and when to do reauth

### LINKED ISSUE #3540: fetch-tool fails and incorrectly retries when cached token is expired
Fluid-fetch tool throw me this error 3 times before giving up.
Re-running it with --forceTokenReauth fixed the problem, but that's bad user experience - it should detect expired token and do right things automatically.
Also in general, 400 is non-retryable code (which should be properly indicated by canRetry property on error object)  -we should not retry (even 3 times). Other (retryable) errors need retries.

Hit error while downloading ops. Retrying
Error: Unable to get access token.
    at Object.createErrorFromResponse (E:\FluidFramework\packages\utils\odsp-utils\dist\odspRequest.js:48:19)
    at fetchTokens (E:\FluidFramework\packages\utils\odsp-utils\dist\odspAuth.js:43:29)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async refreshTokens (E:\FluidFramework\packages\utils\odsp-utils\dist\odspAuth.js:64:23)
    at async Object.refreshTokenFn (E:\FluidFramework\packages\utils\odsp-utils\dist\odspAuth.js:27:5)
    at async Function.timedExecAsync (E:\FluidFramework\packages\utils\telemetry-utils\dist\logger.js:378:25)
    at async E:\FluidFramework\packages\drivers\odsp-driver\dist\vroom.js:23:23 {
  requestResult: {
    href: 'https://login.microsoftonline.com/organizations/oauth2/v2.0/token',
    status: 400,
    data: {
      error: 'invalid_grant',
      error_description: 'AADSTS700082: The refresh token has expired due to inactivity. The token was issued on 2020-04-09T20:43:30.4835848Z and was inactive for 90.00:00:00.\r\n' +
        'Trace ID: f1dd6095-2339-4a56-b850-508508f10000\r\n' +
        'Correlation ID: f72f62b1-385e-4d57-b320-a8f0a6ec1dc8\r\n' +
        'Timestamp: 2020-09-05 01:06:05Z',
      error_codes: [Array],
      timestamp: '2020-09-05 01:06:05Z',
      trace_id: 'f1dd6095-2339-4a56-b850-508508f10000',
      correlation_id: 'f72f62b1-385e-4d57-b320-a8f0a6ec1dc8',
      error_uri: 'https://login.microsoftonline.com/error?code=700082'
    }
  }
}
ERROR: Error: Giving up after 3 attempts to download chunk.
    at loadChunk (E:\FluidFramework\packages\tools\fetch-tool\dist\fluidFetchMessages.js:48:11)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)

### PR DESCRIPTION
**Pull request by jatgarg**
Fixes: https://github.com/microsoft/FluidFramework/issues/3540
1.) Rename odsp-utils -> odsp-doclib-utils.
2.) Allow odsp-doclib-utils to take dep on driver utils and driver deninitions.
3.) Dedup the code between utils and driver.
4.) Wrong property was read on the error which was created from createErrorFromResponse and because of that we were not going in the reauth flow when the refresh token was expired.

### CODE REVIEW

File: packages/tools/fetch-tool/src/fluidFetchSharePoint.ts [hunk 1]
```diff
@@ -56,14 +56,11 @@ export async function resolveWrapper<T>(
...
         return result;
     } catch (e) {
-        if (e.requestResultError) {
-            const parsedBody = JSON.parse(e.requestResult.data);
-            if (parsedBody.error === "invalid_grant"
-                && parsedBody.suberror === "consent_required"
-                && !forceTokenReauth
-            ) {
+        if (e.requestResult) {
+            const parsedBody = e.requestResult.data;
```
Comment (vladsud) [inline packages/tools/fetch-tool/src/fluidFetchSharePoint.ts:60]: At a minimum, I'd export proper interface from odsp driver for this code to leverage, and thus be in sync with any changes happening in the driver.

I did not spend that much time examining code, but it feels like createErrorFromResponse() and code around it creates a parallel error handling system, or rather leaks particular driver-specific way of dealing with error handling to outside world.
I'd try to either map these errors to DriverErrorType (generic), OdspErrorType (odsp specific) or maybe even create new enum, but have mapping of various interesting cases into a flat list of known and expected error codes.

Curious, why shape of errors and how we treat them is different on this code path compared to shipping code paths (see createOdspNetworkError())?

Comment (jatgarg) [reply packages/tools/fetch-tool/src/fluidFetchSharePoint.ts:60]: 1.) createErrorFromResponse() is used to create/throw errors in odsp utils package. That error is getting parsed here for doing reauth. The other errors are just thrown off again.
2.) The error that is coming here is not either the DriverErrorType or OdspErrorType but it is the RequestResultError type which is the error type used in the get/put request in odsp utils package to do some auths etc outside of driver code.

Comment (vladsud) [reply packages/tools/fetch-tool/src/fluidFetchSharePoint.ts:60]: I think we need more type safety. I'd do at least minimal part here by clearly articulating expected interface for errors.
All the other points you bring can be easily addressed as well (by moving code around from drive to utils package, or just defining new enum and simply using same structure for error reporting, i.e. we can leave driver as is and just make sure that there is some reasonable translation of errors happens in utils package that formats errors in the same way as the driver).

Comment (jatgarg) [reply packages/tools/fetch-tool/src/fluidFetchSharePoint.ts]: Added odspNetworkError to odsp utils and removed parallel createErrorFromResponse. Also added proper parsing of auth error.

Comment (jatgarg) [reply packages/tools/fetch-tool/src/fluidFetchSharePoint.ts]: Added the enums and types here only as the utils cannot depend on driver utils

File: packages/utils/odsp-utils/src/odspDrives.ts [hunk 2]
```diff
@@ -120,7 +117,7 @@ export async function getChildrenByDriveItem(
     do {
         const getChildrenResult = await getAsync(url, authRequestInfo);
         if (getChildrenResult.status !== 200) {
-            throw createErrorFromResponse("Unable to get children", getChildrenResult);
+            throw throwOdspNetworkError("Unable to get children", getChildrenResult);
```
Comment (vladsud) [inline packages/utils/odsp-utils/src/odspDrives.ts:120]: 'throw throwOdspNetworkError' does not make sense - only one throw should be on the line :)

File: packages/utils/odsp-utils/src/odspErrorUtils.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { AxiosResponse } from "axios";
+
+export const offlineFetchFailureStatusCode: number = 709;
```
Comment (vladsud) [inline packages/utils/odsp-utils/src/odspErrorUtils.ts:8]: This is all duplicated code, right? Can you please dedup it? 
I was wrong that odsp-driver uses odsp-utils, but I think nothing prevents from establishing this link and driver consuming all this code (minus axios specifics).

Unrelated to your changes - but do you know why we use axios? I.e. why exactly same code using fetch() is not used here? Is it' just historical thing, or there are some good reasons?

File: packages/utils/odsp-utils/src/odspDrives.ts (hunk 2 above)
Comment (jatgarg) [reply packages/utils/odsp-utils/src/odspDrives.ts]: fixed

File: packages/utils/odsp-utils/src/odspErrorUtils.ts (hunk 3 above)
Comment (jatgarg) [reply packages/utils/odsp-utils/src/odspErrorUtils.ts]: I deduped the code. Odsp drive now uses utils code.
Renamed the package from odsp-utils to odsp-doclib-utils.

For axios, I will do that in a separate PR. This change is already big.

**Comment by jatgarg** [comment]
Ping

**Comment by jatgarg** [comment]
Ping

File: packages/utils/odsp-utils/src/odspErrorUtils.ts (hunk 3 above)
Comment (heliocliu) [reply packages/utils/odsp-utils/src/odspErrorUtils.ts]: Does the package really need to be renamed?  It seems unnecessary and the current naming doesn't seem "wrong" now that you're consuming it as expected

Comment (jatgarg) [reply packages/utils/odsp-utils/src/odspErrorUtils.ts]: Actually we(@curtisman and me) wanted to rename it to make sure that it is not a general dumping ground for odsp utils which exists within the odsp driver. So we renamed it to odsp-doclib-utils.

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts [hunk 4]
```diff
@@ -15,11 +15,12 @@ import {
     ISequencedDocumentMessage,
     ISignalMessage,
 } from "@fluidframework/protocol-definitions";
+import { OdspError } from "@fluidframework/odsp-doclib-utils";
```
Comment (heliocliu) [inline packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:18]: alphabetize package import

File: packages/tools/fetch-tool/src/fluidFetchSharePoint.ts [hunk 5]
```diff
@@ -56,19 +57,11 @@ export async function resolveWrapper<T>(
...
-            if (parsedBody.error === "invalid_grant"
-                && parsedBody.suberror === "consent_required"
-                && !forceTokenReauth
-            ) {
-                // Re-auth
-                return resolveWrapper<T>(callback, server, clientConfig, true);
-            }
-            const responseMsg = JSON.stringify(parsedBody.error, undefined, 2);
-            return Promise.reject(`Fail to connect to ODSP server\nError Response:\n${responseMsg}`);
+        if (e.errorType === DriverErrorType.authorizationError && !forceTokenReauth) {
```
Comment (heliocliu) [inline packages/tools/fetch-tool/src/fluidFetchSharePoint.ts:60]: is this right? `!forceTokenReauth` leads to re-auth?

File: packages/tools/webpack-fluid-loader/src/odspUrlResolver.ts [hunk 6]
```diff
@@ -9,7 +9,7 @@ import { OdspDriverUrlResolver, createOdspUrl } from "@fluidframework/odsp-drive
 import {
     IOdspAuthRequestInfo,
     getDriveItemByRootFileName,
-} from "@fluidframework/odsp-utils";
+} from "@fluidframework/odsp-doclib-utils";
```
Comment (heliocliu) [inline packages/tools/webpack-fluid-loader/src/odspUrlResolver.ts:12]: alphabetize, should be before odsp-driver now

File: packages/utils/odsp-doclib-utils/src/odspAuth.ts [hunk 7]
```diff
@@ -3,9 +3,10 @@
  * Licensed under the MIT License.
  */
 
-import { AxiosRequestConfig } from "axios";
-import { IRequestResult, createErrorFromResponse, unauthPostAsync } from "./odspRequest";
-import { getSharepointTenant } from "./odspUtils";
+import { AxiosRequestConfig, AxiosResponse } from "axios";
+import { throwOdspNetworkError } from "./odspErrorUtils";
+import { unauthPostAsync } from "./odspRequest";
+import { getSharepointTenant } from "./odspDocLibUtils";
```
Comment (heliocliu) [inline packages/utils/odsp-doclib-utils/src/odspAuth.ts:9]: alphabetize

File: packages/tools/fetch-tool/src/fluidFetchSharePoint.ts (hunk 5 above)
Comment (jatgarg) [reply packages/tools/fetch-tool/src/fluidFetchSharePoint.ts:60]: Yes this means that if forceTokenReauth was true then it should have been already done. No need to do again even on error.

File: packages/utils/odsp-doclib-utils/src/odspAuth.ts (hunk 7 above)
Comment (jatgarg) [reply packages/utils/odsp-doclib-utils/src/odspAuth.ts:9]: Cool, I will alphabetize in all files.

File: tools/build-tools/src/bumpVersion/README.md [hunk 8]
```diff
@@ -83,7 +83,7 @@ Creating release 0.16.1
     Waiting for package to publish @fluid-example/canvas@0.16.1...
 ...cut for brevity...
     Waiting for package to publish @fluidframework/tool-utils@0.16.1...
-    Waiting for package to publish @fluidframework/odsp-utils@0.16.1...
+    Waiting for package to publish @fluidframework/odsp-doclib-utils@0.16.1...
```
Comment (heliocliu) [inline tools/build-tools/src/bumpVersion/README.md:86]: this one probably shouldn't be changed

**Review by heliocliu — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts (hunk 4 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts]: done

File: packages/tools/webpack-fluid-loader/src/odspUrlResolver.ts (hunk 6 above)
Comment (jatgarg) [reply packages/tools/webpack-fluid-loader/src/odspUrlResolver.ts:11]: done

File: tools/build-tools/src/bumpVersion/README.md (hunk 8 above)
Comment (jatgarg) [reply tools/build-tools/src/bumpVersion/README.md:86]: because it is pointing to version when it was of other name?

Comment (jatgarg) [reply tools/build-tools/src/bumpVersion/README.md]: hmm, makes sense

**Pull request merged**